File: SMEOihaneSaez/R/discretizacion.R
#' Funcion de discretizacion Equal Width
#'
#' @description Esta funcion discretiza un vector numerico en intervalos de longitud iguales
#' @param x será un vector de clase numeric
#' @param num.bins es un numero que indica cuantos intervalos queremos de clase numeric
#' @return Un vector de clase factor
#' @examples
#' a <- c(100, 50, 60, 10, 80, 20)
#' discretizeEW(a, 4)
discretizeEW <- function(x, num.bins) {
  # Encontrar el valor mínimo y máximo de x
  num_max <- max(x)
  num_min <- min(x)

  # Calcular el tamaño del intervalo
  length_lim <- (num_max - num_min) / num.bins

  # Crear los puntos de corte
  puntos_de_corte <- num_min + (1:(num.bins - 1)) * length_lim

  # Inicializar el vector categórico
  vector_categ <- character(length(x))

  # Asignar categorías
  for (i in 1:length(x)) {
    if (x[i] <= min(puntos_de_corte)) {
      vector_categ[i] <- paste0("I(-infty,", min(puntos_de_corte), "]")
    } else if (x[i] >= max(puntos_de_corte)) {
      vector_categ[i] <- paste0("I[",max(puntos_de_corte), ", +infty)")
    } else {
      # Aquí se ajusta el bucle para incluir el punto de corte
      for (j in 2:(num.bins - 1)) {
        if (x[i] <= puntos_de_corte[j]) {
          vector_categ[i] <- paste0("I(", round(puntos_de_corte[j - 1], 2), ",", round(puntos_de_corte[j], 2), "]")
          break
        }
      }
    }
  }

  # Devuelve el vector categorico
  vector_categ<- as.factor(vector_categ)
  return(vector_categ)
}

#' Funcion de discretizacion Equal Width para cada columna de un data frame
#'
#' @description Esta funcion discretiza las columnas numericas de un data frame en intervalos de longitud iguales
#' @param x será un data frame de clase numeric
#' @param num.bins es un numero que indica cuantos intervalos queremos de clase numeric
#' @return Un data frame de clase factor
#' @examples
#' data <- data.frame(Variable1 = c(2.5, 3.6, 5.1, 8.4, 9.7, 2.3, 7.5, 6.2),
#' Variable2 = c(15.2, 18.4, 20.1, 14.5, 12.3, 17.6, 21.8, 19.2))
#' discretizeEW_df(data, num.bins = 3)
discretizeEW_df <- function(x, num.bins) {
  df_dis<- as.data.frame(apply(x, num.bins, MARGIN=2, FUN=discretizeEW))
  return(df_dis)
}


#' Funcion de discretizacion Equal Frecuency
#'
#' @description Esta funcion discretiza un vector numerico en intervalos que contengan el mismo numero de elementos
#' @param x será un vector de clase numeric
#' @param num.bins es un numero que indica cuantos intervalos queremos de clase numeric
#' @return Un vector de clase factor
#' @examples
#' a <- c(100, 50, 60, 10, 80, 20)
#' discretizeEF(a, 4)
discretizeEF <- function(x, num.bins) {

  #calcular el número de valores en cada intervalo
  num_val<- ceiling(length(x)/num.bins) #ceiling() para redondear el número al valor más grande
  x_orden <- sort(x) #ordeno los valores de menor a mayor

  # Inicializar el vector categórico
  vector_categ <- character(length(x))

  # Conseguir puntos de corte
  puntos_de_corte <- c()
  for (i in seq(num_val, length(x), by = num_val)) {
    if (length(puntos_de_corte)==num.bins-1){
      break
    }
    else {
      puntos_de_corte <- c(puntos_de_corte, x_orden[i])
    }
  }
  # Asignar categorías
  for (i in 1:length(x)) {
    if (x[i] <= min(puntos_de_corte)) {
      vector_categ[i] <- paste0("I(-infty,", min(puntos_de_corte), "]")
    } else if (x[i] > max(puntos_de_corte)) {
      vector_categ[i] <- paste0("I[",max(puntos_de_corte), ", +infty)")
    } else {
      # Aquí se ajusta el bucle para incluir el punto de corte
      for (j in 2:(num.bins - 1)) {
        if (x[i] <= puntos_de_corte[j]) {
          vector_categ[i] <- paste0("I(", round(puntos_de_corte[j - 1], 2), ",", round(puntos_de_corte[j], 2), "]")
          break
        }
      }
    }
  }

  # Devuelve vector categorico
  vector_categ<- as.factor(vector_categ)
  return(vector_categ)
}

#' Funcion de discretizacion Equal Frecuency para cada columna de un data frame
#'
#' @description Esta funcion discretiza las columnas numericas de un data frame en intervalos que contengan el mismo numero de elementos
#' @param x será un data frame de clase numeric
#' @param num.bins es un numero que indica cuantos intervalos queremos de clase numeric
#' @return Un data frame de clase factor
#' @examples
#' data <- data.frame(Variable1 = c(2.5, 3.6, 5.1, 8.4, 9.7, 2.3, 7.5, 6.2),
#' Variable2 = c(15.2, 18.4, 20.1, 14.5, 12.3, 17.6, 21.8, 19.2))
#' discretizeEF_df(data, num.bins = 3)
discretizeEF_df <- function(x, num.bins) {
  df_dis<- as.data.frame(apply(x, num.bins, MARGIN=2, FUN=discretizeEF))
  return(df_dis)
}

#' Funcion de discretizacion dado los puntos de corte
#'
#' @description Esta funcion discretiza un vector numerico en intervalos utilizando los puntos de corte dados
#' @param x será un vector de clase numeric
#' @param cut.points es una lista de puntos de corte de clase numeric
#' @return Un vector de clase factor
#' @examples
#' a <- c(100, 50, 60, 10, 80, 20)
#' puntoscorte<- list(32.5, 55.0, 77.5)
#' discretize(a, puntoscorte)
discretize <- function(x, cut.points) {
  puntos_de_corte <- as.numeric(unlist(cut.points))

  # Inicializar el vector categórico
  vector_categ <- character(length(x))

  # Asignar categorías
  for (i in 1:length(x)) {
    if (x[i] <= min(puntos_de_corte)) {
      vector_categ[i] <- paste0("I(-infty,", min(puntos_de_corte), "]")
    } else if (x[i] >= max(puntos_de_corte)) {
      vector_categ[i] <- paste0("I[",max(puntos_de_corte), ", +infty)")
    } else {
      # Aquí se ajusta el bucle para incluir el punto de corte
      for (j in 2:length(puntos_de_corte)) {
        if (x[i] <= puntos_de_corte[j]) {
          vector_categ[i] <- paste0("I(", round(puntos_de_corte[j - 1], 2), ",", round(puntos_de_corte[j], 2), "]")
          break
        }
      }
    }
  }

  vector_categ<- as.factor(vector_categ)
  return(vector_categ)

}

File: SMEOihaneSaez_py/SMEOihaneSaez_py/discretizacion.py
import numpy as np
import pandas as pd

def discretize_ew(x, num_bins):
    """
    Funcion que discretiza un vector numerico en intervalos de longitud iguales

    Parameters
    ----------
    x (float array): un vector numerico a discretizar
    num_bins (int) : el valor del número de intervalos que queremos

    Returns
    -------
    vector_categ (pd.Categorical): un vector de tipo categórico con los elementos del vector numerico discretizados
    
    Example:
        # Definir el vector de entrada
        a = np.array([100, 50, 60, 10, 80, 20])

        # Llamar a la función
        result = discretize_ew(a, 4)

        # Mostrar el resultado
        print(result)

    """
    # Encontrar el valor mínimo y máximo de x
    num_max = np.max(x)
    num_min = np.min(x)

    # Calcular el tamaño del intervalo
    length_lim = (num_max - num_min) / num_bins

    # Crear los puntos de corte
    puntos_de_corte = num_min + np.arange(1, num_bins) * length_lim

    # Inicializar el vector categórico
    vector_categ = []

    # Asignar categorías
    for value in x:
        if value <= puntos_de_corte[0]:
            vector_categ.append(f"I(-infty, {puntos_de_corte[0]:.2f}]")
        elif value >= puntos_de_corte[-1]:
            vector_categ.append(f"I[{puntos_de_corte[-1]:.2f}, +infty)")
        else:
            # Ajuste para incluir el punto de corte adecuado
            for j in range(1, len(puntos_de_corte)):
                if value <= puntos_de_corte[j]:
                    vector_categ.append(f"I({puntos_de_corte[j - 1]:.2f}, {puntos_de_corte[j]:.2f}]")
                    break

    # Convertir a factor/categoría
    vector_categ = pd.Categorical(vector_categ)
    return vector_categ


def discretize_ew_df(df, num_bins):
    """
    Funcion de discretizacion Equal Width para cada columna de un data frame

    Parameters
    ----------
    df (pd.DataFrame): data frame de clase numérico
    num_bins (int): el valor del número de intervalos que queremos

    Returns
    -------
    df_dis (pd.DataFrame): data frame de tipo categórico con los valores discretizados
    
    Example:
        # Crear el DataFrame
        data = pd.DataFrame({
            'Variable1': [2.5, 3.6, 5.1, 8.4, 9.7, 2.3, 7.5, 6.2],
            'Variable2': [15.2, 18.4, 20.1, 14.5, 12.3, 17.6, 21.8, 19.2]
        })

        # Llamar a la función para discretizar el DataFrame
        result_df = discretize_ew_df(data, num_bins=3)

        # Mostrar el resultado
        print(result_df)

    """
    # Aplicar la función discretize_ew a cada columna del DataFrame
    df_dis = df.apply(lambda x: discretize_ew(x, num_bins))
    return df_dis


def discretize_ef(x, num_bins):
    """
    Esta funcion discretiza un vector numerico en intervalos que contengan el mismo numero de elementos

    Parameters
    ----------
    x (array, float): es un vector numérico a discretizar
    num_bins (int): el valor del número de intervalos que queremos

    Returns
    -------
    vector_categ (array, pd.Categorical): es un vector categórico con los valores discretizados.
    
    Example:
        # Definir el vector de entrada
        a = np.array([100, 50, 60, 10, 80, 20])

        # Llamar a la función
        result = discretize_ef(a, 3)

        # Mostrar el resultado
        print(result)
    """
    # Calcular el número de valores en cada intervalo
    num_val = int(np.ceil(len(x) / num_bins))  # ceiling para redondear al valor más alto
    x_sorted = np.sort(x)  # Ordenar los valores de menor a mayor

    # Inicializar el vector categórico
    vector_categ = [''] * len(x)

    # Conseguir puntos de corte
    puntos_de_corte = []
    for i in range(num_val - 1, len(x_sorted), num_val):
        if len(puntos_de_corte) == num_bins - 1:
            break
        else:
            puntos_de_corte.append(x_sorted[i])

    # Asignar categorías
    for i in range(len(x)):
        if x[i] <= min(puntos_de_corte):
            vector_categ[i] = f"I(-infty, {min(puntos_de_corte):.2f}]"
        elif x[i] > max(puntos_de_corte):
            vector_categ[i] = f"I[{max(puntos_de_corte):.2f}, +infty)"
        else:
            # Ajustar el bucle para incluir el punto de corte
            for j in range(1, len(puntos_de_corte)):
                if x[i] <= puntos_de_corte[j]:
                    vector_categ[i] = f"I({puntos_de_corte[j - 1]:.2f}, {puntos_de_corte[j]:.2f}]"
                    break

    # Convertir a factor/categoría
    vector_categ = pd.Categorical(vector_categ)
    return vector_categ


def discretize_ef_df(df, num_bins):
    """
    Funcion de discretizacion Equal Frecuency para cada columna de un data frame

     Parameters
     ----------
     df (pd.DataFrame): data frame de clase numérico
     num_bins (int): el valor del número de intervalos que queremos

     Returns
     -------
     df_dis (pd.DataFrame): data frame de tipo categórico con los valores discretizados
     
     Example:
        # Crear el DataFrame
        data = pd.DataFrame({
            'Variable1': [2.5, 3.6, 5.1, 8.4, 9.7, 2.3, 7.5, 6.2],
            'Variable2': [15.2, 18.4, 20.1, 14.5, 12.3, 17.6, 21.8, 19.2]
        })

        # Llamar a la función para discretizar el DataFrame
        result_df = discretize_ef_df(data, num_bins=3)

        # Mostrar el resultado
        print(result_df) 

    """
    # Aplicar la función discretize_ef a cada columna del DataFrame
    df_dis = df.apply(lambda col: discretize_ef(col, num_bins))
    return df_dis


def discretize(x, cut_points):
    """
    Funcion de discretizacion dado los puntos de corte

    Parameters
    ----------
    x (array, float): un vector de tipo numérico a discretizar 
    cut_points (list): una lista con valores de los puntos de cortes que se deben utilizar para la discretización

    Returns
    -------
    vector_categ (array, pd.Categorical): un vector categórico  con los calores discretizados
    
    Example:
        # Datos de entrada
        a = np.array([100, 50, 60, 10, 80, 20])
        puntoscorte = [32.5, 55.0, 77.5]  # Lista de puntos de corte

        # Llamar a la función discretize
        result = discretize(a, puntoscorte)

        # Mostrar el resultado
        print(result)
    """
    # Convertir los puntos de corte a un array numérico
    puntos_de_corte = np.array(cut_points, dtype=float)

    # Inicializar el vector categórico
    vector_categ = [''] * len(x)

    # Asignar categorías
    for i in range(len(x)):
        if x[i] <= min(puntos_de_corte):
            vector_categ[i] = f"I(-infty, {min(puntos_de_corte):.2f}]"
        elif x[i] >= max(puntos_de_corte):
            vector_categ[i] = f"I[{max(puntos_de_corte):.2f}, +infty)"
        else:
            # Ajustar el bucle para incluir el punto de corte
            for j in range(1, len(puntos_de_corte)):
                if x[i] <= puntos_de_corte[j]:
                    vector_categ[i] = f"I({puntos_de_corte[j - 1]:.2f}, {puntos_de_corte[j]:.2f}]"
                    break

    # Convertir a tipo categórico de pandas
    vector_categ = pd.Categorical(vector_categ)
    return vector_categ

